### PR TITLE
[Baseline] fl-dsme-iot: Federated Learning over IEEE 802.15.4e DSME IoT Networks

### PR DESCRIPTION
**Pull request by SonaliAnand24**
## Summary
Adds a new Flower Baseline simulating federated learning over an IEEE 802.15.4e DSME cluster-tree IoT network with MAC-layer energy and bandwidth constraints derived from two published papers.

## Novel Contribution
First Flower Baseline to incorporate a physical MAC-layer model for FL client participation. Each client is an IoT end device whose:
- Training is gated by DSME energy budget (PSO-DSME paper, Eq. 3)
- Gradient upload is constrained by GTS slot availability (SeCAP paper)
- Energy cost is halved when SeCAP NCR mode activates (61% saving)

## Key Results (20 rounds, CIFAR-10, 10 clients)
- Train loss: 2.18 → 1.97
- Eval accuracy: 11% → 28% (peak round 15), final 25.8%
- SeCAP NCR mode: 20.1 mJ vs 51.8 mJ CR mode (61% energy saving)
- DSMEFedAvg handles all-depleted rounds gracefully

## References
- Anand et al. (2024) PSO-DSME: power model Eq. 3
- Anand et al. (2025) SeCAP: bandwidth model and NCR mode

## How to Run
pip install -e .
flwr run .

## Checklist
- [x] README with setup, results, environment, citations
- [x] All hyperparameters in pyproject.toml
- [x] flwr run . verified on Apple M1, Python 3.12, Flower 1.31
- [x] Novel contribution clearly stated

### CODE REVIEW

File: baselines/fldsme/fldsme/model.py [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+            images = batch["img"]
+            labels = batch["label"]
+            optimizer.zero_grad()
+            loss = criterion(net(images.to(device)), labels.to(device))
+            loss.backward()
+            optimizer.step()
+            running_loss += loss.item()
+
+    avg_trainloss = running_loss / len(trainloader)
+    return avg_trainloss
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/model.py:66]: Good catch — fixed in the latest commit. Divided by total batch count across all epochs rather than len(trainloader)

File: baselines/fl-dsme-iot/pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+[tool.ruff.lint]
+select = ["D", "E", "F", "W", "B", "ISC", "C4", "UP"]
+fixable = ["D", "E", "F", "W", "B", "ISC", "C4", "UP"]
+ignore = ["B024", "B027", "D205", "D209"]
+
+[tool.ruff.lint.pydocstyle]
+convention = "numpy"
+
+[tool.hatch.build.targets.wheel]
+packages = ["fl_dsme_iot"]
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fl-dsme-iot/pyproject.toml]: Fixed — renamed the baseline directory and Python package to fldsme to match Flower's naming convention (following fedprox, fedavgm, fednova etc.).
 Structure is now baselines/fldsme/fldsme/ 
so the baseline tooling script should pass.

File: baselines/fldsme/fldsme/dsme_model.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+        seed: int = 42,
+    ):
+        self.bo = bo
+        self.mo = mo
+        self.so = so
+        self.num_clients = num_clients
+        self.num_clusters = num_clusters
+        self.energy_budget = energy_budget
+        self.bandwidth_frac = bandwidth_frac
+        self._rng = np.random.default_rng(seed)
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/dsme_model.py]: Fixed — removed self._rng and the unused seed parameter. 
Randomness in effective_bandwidth_fraction is deterministic per client+round via np.random.default_rng(client_id * 1000 + fl_round), which is clearer and documented

File: baselines/fldsme/fldsme/server_app.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    fraction_train: float = float(context.run_config["fraction-train"])
+
+    global_model = Net()
+    arrays = ArrayRecord(global_model.state_dict())
+
+    strategy = DSMEFedAvg(
+        fraction_train=fraction_train,
+        fraction_evaluate=1.0,
+        min_available_nodes=2,
+    )
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/server_app.py:280]: Fixed with a custom train_metrics_aggr_fn (dsme_metrics_aggregation). 
It now separately tracks active_clients, skipped_clients, and total_sampled per round, with weighted averages computed only over active clients. 
Skipped clients are counted explicitly rather than being silently zeroed out by the num-examples weighting.

File: baselines/fldsme/fldsme/server_app.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+                    pass
+
+        if total_examples == 0:
+            # All clients energy-depleted: hold global model unchanged.
+            # Models a DSME beacon interval with no successful transmissions.
+            print(
+                f"\n[Round {server_round}] All sampled clients energy-depleted."
+                " Keeping current global model (DSME beacon interval skipped).\n"
+            )
+            return None, None
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/server_app.py]: Fixed — the all-depleted branch now returns None for arrays (global model held unchanged), but returns a MetricRecord with skipped_clients=n_sampled, total_sampled=n_sampled, and active_clients=0.0. 
Skipped rounds now appear in the experiment logs with correct DSME participation statistics instead of being silently dropped. 
See commit 7f59d1795.

File: baselines/fldsme/fldsme/client_app.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+    state_dict = model.state_dict()
+    masked = {}
+    for key, tensor in state_dict.items():
+        flat = tensor.flatten().clone()
+        n_keep = max(1, int(len(flat) * bw_frac))
+        if n_keep < len(flat):
+            _, idx = torch.topk(flat.abs(), n_keep)
+            mask = torch.zeros_like(flat)
+            mask[idx] = 1.0
+            masked[key] = (flat * mask).reshape(tensor.shape)
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/client_app.py]: Fixed — the bandwidth mask now operates on the update delta (trained - global) rather than the absolute weights. 
We compute sparse_update = top-k(trained - global), then reconstruct as global + sparse_update before returning the ArrayRecord. 
FedAvg now aggregates valid compressed updates instead of corrupted zeroed weights. See commit

File: baselines/fldsme/fldsme/server_app.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    )
+
+    result = strategy.start(
+        grid=grid,
+        initial_arrays=arrays,
+        num_rounds=num_rounds,
+    )
+
+    print("\nSaving final model to disk...")
+    state_dict = result.arrays.to_torch_state_dict()
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/server_app.py]: Fixed — we now keep a reference to initial_state_dict before strategy.start(). 
After training, we check result.arrays is not None and len(result.arrays) > 0 before calling to_torch_state_dict(). 
If all rounds were skipped and result.arrays is empty, we fall back to the initial global model, ensuring final_model.pt always contains a valid model.

File: baselines/fldsme/fldsme/client_app.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    # We sparsify the UPDATE (trained - global), not the absolute weights.
+    # Zeroing absolute weights would corrupt the model (P1 bot review fix).
+    # Reconstruct as: global_weights + sparse_update for FedAvg aggregation.
+    global_state = arrays.to_torch_state_dict()   # received global weights
+    trained_state = model.state_dict()             # locally trained weights
+    masked = {}
+    for key in trained_state:
+        global_tensor  = global_state[key].float()
+        trained_tensor = trained_state[key].float()
+        update = (trained_tensor - global_tensor).flatten().clone()
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/client_app.py]: Fixed — added .to(trained_tensor.device) when loading global_tensor so both tensors are on the same device before the subtraction. arrays.to_torch_state_dict() always produces CPU tensors, but train_fn may move the model to CUDA, so the explicit device transfer prevents the mismatch.

File: baselines/fldsme/fldsme/server_app.py [hunk 9]
```diff
@@ -0,0 +1,167 @@
...
+            if not msg.has_error():
+                try:
+                    n = msg.content.metric_records["metrics"].get(
+                        "num-examples", 0
+                    )
+                    total_examples += int(n)
+                except (KeyError, AttributeError):
+                    pass
+
+        if total_examples == 0:
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/server_app.py]: Fixed — we now first separate valid replies (no error, metric_records accessible) from error replies. 
The DSME all-depleted path only activates when len(valid_replies) > 0 and total_examples == 0, meaning at least one client successfully reported num-examples=0 (genuine energy depletion).
 When all replies are errors, we fall through to super().aggregate_train() so FedAvg's _check_and_log_replies logs the failures correctly.

File: baselines/fldsme/fldsme/client_app.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    num_partitions = int(context.node_config["num-partitions"])
+    fl_round = int(msg.content.config_records["config"]["server-round"])
+
+    mac_model = _build_mac_model(context, num_partitions)
+    profile = mac_model.get_client_profile(
+        client_id=partition_id,
+        fl_round=fl_round,
+        model_size_kb=MODEL_SIZE_KB,
+    )
+    eligible = mac_model.is_eligible(profile, model_size_kb=MODEL_SIZE_KB)
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/client_app.py]: Fixed — local_epochs is now read from context.run_config before the MAC model call and passed explicitly as n_local_epochs=int(local_epochs) into both get_client_profile() and is_eligible(). 
The energy gate and the reported energy_used_mj now use the same epoch count, making the check valid for any configured local-epochs value.

File: baselines/fldsme/fldsme/server_app.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+        # valid reply (i.e. a client successfully reported num-examples=0).
+        # If all replies are errors, fall through to FedAvg which will log
+        # the failures correctly via _check_and_log_replies.
+        if valid_replies:
+            total_examples = sum(
+                int(msg.content.metric_records["metrics"].get("num-examples", 0))
+                for msg in valid_replies
+            )
+
+            if total_examples == 0:
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/fldsme/server_app.py:214]: Fixed — before taking the all-depleted return path, we now explicitly log all error_replies via Flower's WARNING logger so failures are never silently swallowed. 
Each failed reply's error is logged individually; we also add error_replies as a field in the returned MetricRecord so the experiment logs capture both the depletion count and any concurrent failures in the same round.

**Comment by SonaliAnand24** [comment]
Updated after bot review. Three things changed:
**Added a control arm**. dsme-enabled=false runs plain FedAvg through the identical code path via NullMACModel. The MAC-layer constraints cost 6.4 accuracy points (31.2% → 24.8%, 5 seeds).

**Fixed seeding**. The bandwidth RNG ignored the run seed and the global model init was unseeded, so multi-seed runs weren't independent. seed=0 reproduces the old draws exactly.

**Regenerated all results**. The figures in the original description predate f41380a, the delta-sparsification fix — they describe a run whose aggregation was corrupting the global model. Everything in the README is post-fix, 5 seeds, with error bars.

@jafermarq, @chongshenng, @tanertopal, @danieljanes — Please do review when you have time.

File: baselines/fldsme/run_experiments.py [hunk 12]
```diff
@@ -0,0 +1,194 @@
...
+    },
+    # Current PR behaviour: MAC gate active, uniform random sampling.
+    "dsme_random": {
+        "dsme-enabled": True,
+        "selection-policy": "random",
+    },
+    # Energy-aware selection, no harvesting.
+    "dsme_eligible": {
+        "dsme-enabled": True,
+        "selection-policy": "eligible",
```
[the remark this answers is not in the document]
Comment (SonaliAnand24) [reply baselines/fldsme/run_experiments.py]: 👍 DSMEFedAvg inherits FedAvg's uniform configure_train, so those arms were the same code path as dsme_random under a different RNG stream and couldn't measure the policy they named. Dropped the arms and the four config keys (selection-policy, harvest-mj-per-round, energy-margin, exploration-eps) that only an energy-aware strategy would read.
Running them wasn't wasted: five seeds of the duplicate arm against five of dsme_random gives a 0.7-point accuracy spread, which is the run-to-run noise floor now quoted in the README. Energy-aware selection will come as a follow-up PR that actually overrides configure_train.
